extensions/interactions/SetInput/directives/set-input-validation.service.ts

```diff
@@ -34,13 +37,124 @@ export class SetInputValidationService {
     private baseInteractionValidationServiceInstance:
       baseInteractionValidationService) {}
 
+  private isSameSet(inputA: string[], inputB: string[]): boolean {
+    let setA = new Set(inputA);
+    let setB = new Set(inputB);
+    if (setA.size !== setB.size) {
+      return false;
+    }
+    for (let element of setA.values()) {
+      if (!setB.has(element)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private isSubset(inputA: string[], inputB: string[]): boolean {
+    // Checks if set inputA is a subset of inputB.
+    let setB = new Set(inputB);
+    return inputA.every(val => setB.has(val));
+  }
+
+  private isSameRule(ruleA: Rule, ruleB: Rule): boolean {
+    return ruleA.type === ruleB.type &&
+      this.isSameSet(<string[]>ruleA.inputs.x, <string[]>ruleB.inputs.x);
+  }
+
   // TODO(#7176): Replace 'any' with the exact type. This has been kept as
   // 'any' because 'customizationArgs' is a dict with possible underscore_cased
   // keys which give tslint errors against underscore_casing in favor of
   // camelCasing.
   getCustomizationArgsWarnings(customizationArgs: any): any[] {
-    // TODO(juansaba): Implement customization args validations.
-    return [];
+    let warningsList = [];
+
+    let buttonText = customizationArgs.buttonText &&
+      customizationArgs.buttonText.value;
+    if (!angular.isString(buttonText)) {
+      warningsList.push({
+        type: AppConstants.WARNING_TYPES.ERROR,
+        message: 'Button text must be a string.'
+      });
+    } else {
+      if (buttonText.length === 0) {
+        warningsList.push({
+          message: 'Label for this button should not be empty.',
+          type: AppConstants.WARNING_TYPES.ERROR
+        });
+      }
+    }
+
+    return warningsList;
+  }
+
+  getRedundantRuleWarnings(answerGroups: AnswerGroup[]): IWarning[] {
+    let warningsList: IWarning[] = [];
+
+    let previousRules: Array<{
+      answerGroupIndex: number,
+      ruleIndex: number,
+      rule: Rule
+    }> = [];
+
+    for (let [answerGroupIndex, answerGroup] of answerGroups.entries()) {
+      for (let [ruleIndex, rule] of answerGroup.rules.entries()) {
+        for (let prevRule of previousRules) {
+          if (this.isSameRule(prevRule.rule, rule)) {
+            warningsList.push({
+              type: AppConstants.WARNING_TYPES.ERROR,
+              message: `Rule ${ruleIndex + 1} from answer group ` +
+                `${answerGroupIndex + 1} is the same as ` +
+                `rule ${prevRule.ruleIndex + 1} from ` +
+                `answer group ${prevRule.answerGroupIndex + 1}`
+            });
+          } else if (prevRule.rule.type === rule.type) {
+            // Only consider rules of the same type as the previous rule
+            // of the same tyle might 'cover' the following one.
```

Comment:
I'm not sure if I understand this, could you try to rephrase.

Reply:
I added some more details for this part, hope it makes sense now

Reply:
Yes, I understand it better now, thanks.